### ISSUE TITLE
Undefined variable in where expression

### ISSUE THREAD

**Issue by mfalt** [JuliaLang/julia#33381]
It seems that sometimes, a method might catch calls that it should not. This results in type parameters not being defined. A minimal working example below:

```julia
function f(A::M, lambda::R) where {R <: Real,  M <: AbstractArray{R}}
    println("M: $M")
    println("R: $R")
    return M, R
end

function g(A::M, lambda::R) where {R <: Real,  C<: Union{R, Complex{R}}, M <: AbstractArray{C}}
    println("M: $M")
    println("C: $C")
    println("R: $R")
    return M, C, R
end

M = randn(2,2)
b = randn(2)

f(M, 1.0) # Works as expected
f(M, 1) # As expected: "MethodError: no method matching f(::Array{Float64,2}, ::Int64)"

g(M, 1.0) # Works as expected
g(im*M, 1.0) # Works as expected
g(im*M, 1) # As expected: "MethodError: no method matching g(::Array{Complex{Float64},2}, ::Int64)"

g(M, 1) # For some reason, the method is called and: "UndefVarError: R not defined"
```
the error is thrown on the line `println("R: $R")`

Verified on 
Julia Version 1.2.0 and 1.0.3

**Comment by KristofferC** [comment]
All parameters don't have to be defined for the signature to match, for example:

```jl
julia> foo(x::Union{Float64, T}) where {T} = T

julia> foo(1)
Int64

julia> foo(2.0)
ERROR: UndefVarError: T not defined
Stacktrace:
 [1] foo(::Float64) at ./REPL[11]:1
 [2] top-level scope at REPL[12]:1
```

**Comment by mfalt** [comment]
> All parameters don't have to be defined for the signature to match, for example:

Interesting, I didn't think of this. But at the same time, I don't understand how `C::Union{R, Complex{R}}` can be defined as `Float64` without `R` being defined in this case.

**Comment by JeffBezanson** [comment]
The problem is that R==Real, C==Float64 is a solution that allows the method to match. But R==Union{Int,Float64} is also a possible solution. So the value of `R` is hard to uniquely define, but the method matches since the argument type is definitely a subtype of the signature.

**Comment by mfalt** [comment]
@JeffBezanson Thanks for explaining, I wasn't aware that `R` was allowed to match with non concrete types, such as `Union{Int,Float64}`. It is still a bit unclear to me, what the proper definition would be of a function that allows an array `M` with elements that are either `Complex{R}` or `R`, and a scalar `s` of type `R`, where `R` is concrete.
Consider 3 cases:

```julia
g0(M::AbstractArray{C}, s::R) where {R<:Real, C<:Union{R,Complex{R}}} = R
g0(randn(2), 1.0) # Works
g0(randn(2), 1) # ERROR: UndefVarError: R not defined
```

The documentation states that if `R` appears multiple times, then `R` has to be concrete, however, this does not seem to be strictly enough, I guess the following is equivalent to g0?
```julia
g1(M::AbstractArray{<:Union{R,Complex{R}}}, s::R) where {R<:Real} = R
g1(randn(2), 1.0) # Works
g1(randn(2), 1) # ERROR: UndefVarError: R not defined
```
So it seems that the only way to properly do this is to explicitly write R everywhere
```julia
g2(M::Union{AbstractArray{R},AbstractArray{Complex{R}}}, v::R) where {R<:Real} = R
g2(randn(2), 1.0) # Works
g2(randn(2), 1) # ERROR: MethodError: no method matching g2(::Array{Float64,1}, ::Int64)
```
which seems quite cumbersome. 

So this is my understanding:
* while we enforce the concreteness in the diagonal rule for the parameters in a function definition (g2 above)
* we do not enforce a diagonal rule in a statement like `C<:Union{R, Complex{R}}`

is this correct?

This seems to lead to some quite interesting results:
```julia
julia> Tuple{Float64,Int64} <: Tuple{T,T} where T
false

julia> Tuple{Float64,Int64} <: Tuple{T,<:Union{T,Complex{T}}} where T
true

julia> Tuple{Float64,Float64,Int64} <: Tuple{T,T,<:Union{T,Complex{T}}} where T
false
```

**Comment by JeffBezanson** [comment]
Right; the diagonal rule only applies to occurrences of the variable as a direct parameter of the Tuple type.

`R` being undefined inside the method is not a show-stopper; for example you can use `typeof(s)` instead.

> a function that allows an array `M` with elements that are either `Complex{R}` or `R`

`C<:Union{R,Complex{R}}` does not do that, since it also allows `C == Union{R,Complex{R}}`. So your solution with `Union{AbstractArray{R},AbstractArray{Complex{R}}}` is correct.

**Comment by StefanKarpinski** [comment]
Is there an actual issue here or is this just a matter of explaining what's happening?

**Comment by mfalt** [comment]
After reading the docs (https://docs.julialang.org/en/v1/devdocs/types/#Diagonal-types-1) a few more times, I finally think I understand what is going on here and realize it seems to be the intended behaviour.

However, I have two points that I think could be discussed:
* If it is the behaviour that users expect, or if concreteness should be applied more generally
* If it would be useful to have some syntax to denote that a variable has to be concrete.

For the first point:
```julia
function g2(A::M, x::R) where {R <: Real,  C<: Union{R, Complex{R}}, M <: AbstractArray{C}}
    A[1] = x
end
```
I think the expectation is that "Any x::R that is caught in the function can be stored in A, since A is a container of either R or Complex{R}", however, this is not true since the diagonal rule does not apply
```julia
A = fill(1//2, 2)
x = BigFloat(pi)
g2(A, x) # ERROR: InexactError: ....
```

For the second point:
Although it is possible in the example above to enforce concreteness of `R` by manually expanding the union to `A::Union{AbstractArray{R},AbstractArray{Complex{R}}`.  If `C` (and/or `M`) are large unions, this could(?) lead to an exponential increase in the number of types that has to be manually and explicitly written. However, I am not sure if there are any real cases where this would occur.

Feel free to close the issue if you think I should open those as separate issues or if they are not valid points.

**Comment by JeffBezanson** [comment]
Can probably be closed as a dup of #30363, which asks for explicit concreteness constraints.
